File: gamedev/c-ios/src/gpu_backend_metal.m
#import <Metal/Metal.h>
#import <Foundation/Foundation.h>
#include "gpu_backend.h"
#include <string.h>

// Internal structures
struct gpu_device {
    id<MTLDevice> device;
    id<MTLCommandQueue> command_queue;
    id<MTLLibrary> default_library;
};

struct gpu_texture {
    id<MTLTexture> texture;
    int width;
    int height;
};

struct gpu_readback_buffer {
    id<MTLBuffer> buffer;
    size_t size;
};

struct gpu_command_buffer {
    id<MTLCommandBuffer> cmd_buffer;
    bool completed;
};

struct gpu_pipeline {
    id<MTLRenderPipelineState> pipeline_state;
    id<MTLDepthStencilState> depth_stencil_state;
};

struct gpu_buffer {
    id<MTLBuffer> buffer;
    size_t size;
};

struct gpu_render_encoder {
    id<MTLRenderCommandEncoder> encoder;
};

gpu_device_t* gpu_init(void) {
    gpu_device_t* device = (gpu_device_t*)malloc(sizeof(gpu_device_t));

    // Create Metal device
    device->device = MTLCreateSystemDefaultDevice();
    if (!device->device) {
        free(device);
        return NULL;
    }

    // Create command queue
    device->command_queue = [device->device newCommandQueue];

    // Try to load shader library
    NSError* error = nil;
    device->default_library = nil;

    // Try multiple paths for the shader file
    NSArray* paths = @[@"triangle.metal",
                       @"src/shaders/triangle.metal",
                       @"../../src/shaders/triangle.metal"];

    for (NSString* path in paths) {
        if ([[NSFileManager defaultManager] fileExistsAtPath:path]) {
            NSString* shaderSource = [NSString stringWithContentsOfFile:path
                                                               encoding:NSUTF8StringEncoding
                                                                  error:&error];
            if (shaderSource) {
                device->default_library = [device->device newLibraryWithSource:shaderSource
                                                                       options:nil
                                                                         error:&error];
                if (device->default_library) {
                    NSLog(@"Loaded shader from: %@", path);
                    break;
                }
            }
        }
    }

    if (!device->default_library) {
        // Try to get default library
        device->default_library = [device->device newDefaultLibrary];
        if (!device->default_library) {
            NSLog(@"Warning: No Metal library found");
        }
    }

    return device;
}

void* gpu_get_native_device(gpu_device_t* device) {
    return (__bridge void*)device->device;
}

gpu_texture_t* gpu_create_texture(gpu_device_t* device, int width, int height) {
    gpu_texture_t* texture = (gpu_texture_t*)malloc(sizeof(gpu_texture_t));
    texture->width = width;
    texture->height = height;

    // Create texture descriptor
    MTLTextureDescriptor *texDesc = [MTLTextureDescriptor
        texture2DDescriptorWithPixelFormat:MTLPixelFormatBGRA8Unorm
        width:width
        height:height
        mipmapped:NO];

    texDesc.storageMode = MTLStorageModePrivate;
    texDesc.usage = MTLTextureUsageRenderTarget | MTLTextureUsageShaderRead;

    texture->texture = [device->device newTextureWithDescriptor:texDesc];

    return texture;
}

void* gpu_get_native_texture(gpu_texture_t* texture) {
    return (__bridge void*)texture->texture;
}

gpu_readback_buffer_t* gpu_create_readback_buffer(gpu_device_t* device, size_t size) {
    gpu_readback_buffer_t* buffer = (gpu_readback_buffer_t*)malloc(sizeof(gpu_readback_buffer_t));
    buffer->size = size;

    buffer->buffer = [device->device newBufferWithLength:size
                                     options:MTLResourceStorageModeShared];

    return buffer;
}

gpu_command_buffer_t* gpu_readback_texture_async(
    gpu_device_t* device,
    gpu_texture_t* texture,
    gpu_readback_buffer_t* buffer,
    int width,
    int height
) {
    gpu_command_buffer_t* cmd = (gpu_command_buffer_t*)malloc(sizeof(gpu_command_buffer_t));
    cmd->completed = false;

    // Create command buffer
    cmd->cmd_buffer = [device->command_queue commandBuffer];

    // Create blit encoder
    id<MTLBlitCommandEncoder> blitEncoder = [cmd->cmd_buffer blitCommandEncoder];

    // Setup copy operation
    [blitEncoder copyFromTexture:texture->texture
                     sourceSlice:0
                     sourceLevel:0
                    sourceOrigin:MTLOriginMake(0, 0, 0)
                      sourceSize:MTLSizeMake(width, height, 1)
                        toBuffer:buffer->buffer
               destinationOffset:0
          destinationBytesPerRow:width * 4
        destinationBytesPerImage:width * height * 4];

    [blitEncoder endEncoding];

    // Add completion handler to mark as complete
    [cmd->cmd_buffer addCompletedHandler:^(id<MTLCommandBuffer> cmdBuf) {
        (void)cmdBuf;
        cmd->completed = true;
    }];

    return cmd;
}

void gpu_submit_commands(gpu_command_buffer_t* cmd_buffer, bool wait) {
    [cmd_buffer->cmd_buffer commit];

    if (wait) {
        [cmd_buffer->cmd_buffer waitUntilCompleted];
        cmd_buffer->completed = true;
    }
}

bool gpu_is_readback_complete(gpu_command_buffer_t* cmd_buffer) {
    if (cmd_buffer->completed) return true;

    if (cmd_buffer->cmd_buffer.status == MTLCommandBufferStatusCompleted) {
        cmd_buffer->completed = true;
        return true;
    }
    return false;
}

void* gpu_get_readback_data(gpu_readback_buffer_t* buffer) {
    return buffer->buffer.contents;
}

void gpu_copy_readback_data(gpu_readback_buffer_t* buffer, void* dst, size_t size) {
    void* src = buffer->buffer.contents;
    memcpy(dst, src, size < buffer->size ? size : buffer->size);
}

void gpu_destroy_command_buffer(gpu_command_buffer_t* cmd_buffer) {
    if (cmd_buffer) {
        // ARC will handle the release
        cmd_buffer->cmd_buffer = nil;
        free(cmd_buffer);
    }
}

void gpu_destroy_texture(gpu_texture_t* texture) {
    if (texture) {
        // ARC will handle the release
        texture->texture = nil;
        free(texture);
    }
}

void gpu_destroy_readback_buffer(gpu_readback_buffer_t* buffer) {
    if (buffer) {
        // ARC will handle the release
        buffer->buffer = nil;
        free(buffer);
    }
}

// === Rendering Implementation ===

gpu_pipeline_t* gpu_create_pipeline(
    gpu_device_t* device,
    const char* shader_source,
    const char* vertex_function,
    const char* fragment_function,
    gpu_vertex_layout_t* vertex_layout
) {
    gpu_pipeline_t* pipeline = (gpu_pipeline_t*)malloc(sizeof(gpu_pipeline_t));

    NSError* error = nil;
    id<MTLLibrary> library = device->default_library;

    // If shader source provided, compile it
    if (shader_source) {
        NSString* source = [NSString stringWithUTF8String:shader_source];
        library = [device->device newLibraryWithSource:source options:nil error:&error];
        if (!library) {
            NSLog(@"Failed to compile shader: %@", error);
            free(pipeline);
            return NULL;
        }
    }

    // Get shader functions
    id<MTLFunction> vertexFunc = [library newFunctionWithName:[NSString stringWithUTF8String:vertex_function]];
    id<MTLFunction> fragmentFunc = [library newFunctionWithName:[NSString stringWithUTF8String:fragment_function]];

    if (!vertexFunc) {
        NSLog(@"Failed to find vertex function: %s", vertex_function);
        // List available functions
        NSArray<NSString*>* functionNames = library.functionNames;
        NSLog(@"Available functions in library:");
        for (NSString* name in functionNames) {
            NSLog(@"  - %@", name);
        }
        free(pipeline);
        return NULL;
    }
    if (!fragmentFunc) {
        NSLog(@"Failed to find fragment function: %s", fragment_function);
        free(pipeline);
        return NULL;
    }
    NSLog(@"Found shader functions: %s, %s", vertex_function, fragment_function);

    // Create vertex descriptor
    MTLVertexDescriptor* vertexDesc = [[MTLVertexDescriptor alloc] init];

    for (int i = 0; i < vertex_layout->num_attributes; i++) {
        gpu_vertex_attr_t* attr = &vertex_layout->attributes[i];

        // Set attribute format
        MTLVertexFormat format;
        switch (attr->format) {
            case 0: format = MTLVertexFormatFloat2; break;
            case 1: format = MTLVertexFormatFloat3; break;
            case 2: format = MTLVertexFormatFloat4; break;
            default: format = MTLVertexFormatFloat4; break;
        }

        vertexDesc.attributes[attr->index].format = format;
        vertexDesc.attributes[attr->index].offset = attr->offset;
        vertexDesc.attributes[attr->index].bufferIndex = 0;
    }

    vertexDesc.layouts[0].stride = vertex_layout->stride;
    vertexDesc.layouts[0].stepFunction = MTLVertexStepFunctionPerVertex;

    // Create pipeline descriptor
    MTLRenderPipelineDescriptor* pipelineDesc = [[MTLRenderPipelineDescriptor alloc] init];
    pipelineDesc.vertexFunction = vertexFunc;
    pipelineDesc.fragmentFunction = fragmentFunc;
    pipelineDesc.vertexDescriptor = vertexDesc;
    pipelineDesc.colorAttachments[0].pixelFormat = MTLPixelFormatBGRA8Unorm;

    // Enable blending to ensure colors are written
    pipelineDesc.colorAttachments[0].blendingEnabled = YES;
    pipelineDesc.colorAttachments[0].rgbBlendOperation = MTLBlendOperationAdd;
    pipelineDesc.colorAttachments[0].alphaBlendOperation = MTLBlendOperationAdd;
    pipelineDesc.colorAttachments[0].sourceRGBBlendFactor = MTLBlendFactorOne;
    pipelineDesc.colorAttachments[0].sourceAlphaBlendFactor = MTLBlendFactorOne;
    pipelineDesc.colorAttachments[0].destinationRGBBlendFactor = MTLBlendFactorZero;
    pipelineDesc.colorAttachments[0].destinationAlphaBlendFactor = MTLBlendFactorZero;

    // Create pipeline state
    pipeline->pipeline_state = [device->device newRenderPipelineStateWithDescriptor:pipelineDesc error:&error];
    if (!pipeline->pipeline_state) {
        NSLog(@"Failed to create pipeline state: %@", error);
        free(pipeline);
        return NULL;
    }
    NSLog(@"Pipeline created successfully");

    // Create depth stencil state (no depth testing for now)
    MTLDepthStencilDescriptor* depthDesc = [[MTLDepthStencilDescriptor alloc] init];
    depthDesc.depthCompareFunction = MTLCompareFunctionAlways;
    depthDesc.depthWriteEnabled = NO;
    pipeline->depth_stencil_state = [device->device newDepthStencilStateWithDescriptor:depthDesc];

    return pipeline;
}

gpu_buffer_t* gpu_create_buffer(gpu_device_t* device, const void* data, size_t size) {
    gpu_buffer_t* buffer = (gpu_buffer_t*)malloc(sizeof(gpu_buffer_t));
    buffer->size = size;

    if (data) {
        buffer->buffer = [device->device newBufferWithBytes:data
                                                      length:size
                                                     options:MTLResourceStorageModeShared];
    } else {
        buffer->buffer = [device->device newBufferWithLength:size
                                                      options:MTLResourceStorageModeShared];
    }

    return buffer;
}

gpu_command_buffer_t* gpu_begin_commands(gpu_device_t* device) {
    gpu_command_buffer_t* cmd = (gpu_command_buffer_t*)malloc(sizeof(gpu_command_buffer_t));
    cmd->cmd_buffer = [device->command_queue commandBuffer];
    cmd->completed = false;
    return cmd;
}

gpu_render_encoder_t* gpu_begin_render_pass(
    gpu_command_buffer_t* cmd_buffer,
    gpu_texture_t* target,
    float clear_r, float clear_g, float clear_b, float clear_a
) {
    if (!cmd_buffer || !target || !target->texture) {
        NSLog(@"ERROR: Invalid parameters to gpu_begin_render_pass");
        return NULL;
    }

    gpu_render_encoder_t* encoder = (gpu_render_encoder_t*)malloc(sizeof(gpu_render_encoder_t));

    // Create render pass descriptor
    MTLRenderPassDescriptor* passDesc = [[MTLRenderPassDescriptor alloc] init];
    passDesc.colorAttachments[0].texture = target->texture;
    passDesc.colorAttachments[0].loadAction = MTLLoadActionClear;
    passDesc.colorAttachments[0].clearColor = MTLClearColorMake(clear_r, clear_g, clear_b, clear_a);
    passDesc.colorAttachments[0].storeAction = MTLStoreActionStore;

    // Keep the requested clear color for now (will be black)

    // Create render encoder
    encoder->encoder = [cmd_buffer->cmd_buffer renderCommandEncoderWithDescriptor:passDesc];
    if (!encoder->encoder) {
        NSLog(@"ERROR: Failed to create render command encoder");
        free(encoder);
        return NULL;
    }

    // Set viewport
    MTLViewport viewport = {
        .originX = 0.0,
        .originY = 0.0,
        .width = (double)target->width,
        .height = (double)target->height,
        .znear = 0.0,
        .zfar = 1.0
    };
    [encoder->encoder setViewport:viewport];

    return encoder;
}

void gpu_set_pipeline(gpu_render_encoder_t* encoder, gpu_pipeline_t* pipeline) {
    [encoder->encoder setRenderPipelineState:pipeline->pipeline_state];
    [encoder->encoder setDepthStencilState:pipeline->depth_stencil_state];
    [encoder->encoder setCullMode:MTLCullModeNone];  // Disable culling
    [encoder->encoder setFrontFacingWinding:MTLWindingCounterClockwise];
}

void gpu_set_vertex_buffer(gpu_render_encoder_t* encoder, gpu_buffer_t* buffer, int index) {
    if (!buffer || !buffer->buffer) {
        NSLog(@"ERROR: Invalid buffer in gpu_set_vertex_buffer");
        return;
    }
    [encoder->encoder setVertexBuffer:buffer->buffer offset:0 atIndex:index];
}

void gpu_set_uniforms(gpu_render_encoder_t* encoder, int index, const void* data, size_t size) {
    [encoder->encoder setVertexBytes:data length:size atIndex:index];
}

void gpu_draw(gpu_render_encoder_t* encoder, int vertex_count) {
    if (!encoder || !encoder->encoder) {
        NSLog(@"ERROR: Invalid encoder in gpu_draw");
        return;
    }
    [encoder->encoder drawPrimitives:MTLPrimitiveTypeTriangle
                          vertexStart:0
                          vertexCount:vertex_count];
}

void gpu_end_render_pass(gpu_render_encoder_t* encoder) {
    [encoder->encoder endEncoding];
    encoder->encoder = nil;
    free(encoder);
}

void gpu_commit_commands(gpu_command_buffer_t* cmd_buffer, bool wait) {
    // Add completion handler to know when GPU work is done
    __block bool* completed_ptr = &cmd_buffer->completed;
    [cmd_buffer->cmd_buffer addCompletedHandler:^(id<MTLCommandBuffer> buffer) {
        *completed_ptr = true;
    }];

    [cmd_buffer->cmd_buffer commit];
    if (wait) {
        [cmd_buffer->cmd_buffer waitUntilCompleted];
        cmd_buffer->completed = true;
    }
}

void gpu_wait_for_command_buffer(gpu_command_buffer_t* cmd_buffer) {
    if (!cmd_buffer || cmd_buffer->completed) {
        return;
    }

    // Wait for the command buffer to complete
    [cmd_buffer->cmd_buffer waitUntilCompleted];
    cmd_buffer->completed = true;
}

void gpu_destroy_pipeline(gpu_pipeline_t* pipeline) {
    if (pipeline) {
        pipeline->pipeline_state = nil;
        pipeline->depth_stencil_state = nil;
        free(pipeline);
    }
}

void gpu_destroy_buffer(gpu_buffer_t* buffer) {
    if (buffer) {
        buffer->buffer = nil;
        free(buffer);
    }
}

void gpu_destroy(gpu_device_t* device) {
    if (device) {
        // ARC will handle the release
        device->default_library = nil;
        device->command_queue = nil;
        device->device = nil;
        free(device);
    }
}

File: gamedev/c-ios/src/gpu_backend_metal.c
#include "gpu_backend.h"
#include <stdlib.h>
#include <string.h>
#include <Block.h>

// Metal and Foundation headers
#include <objc/objc.h>
#include <objc/runtime.h>
#include <objc/message.h>
#include <Metal/Metal.h>

// Metal forward declarations
typedef struct objc_object *id;
typedef struct objc_selector *SEL;
#ifndef nil
#define nil ((id)0)
#endif

// Metal selectors we'll use
static SEL sel_alloc;
static SEL sel_init;
static SEL sel_release;
static SEL sel_retain;
static SEL sel_newCommandQueue;
static SEL sel_commandBuffer;
static SEL sel_commit;
static SEL sel_waitUntilCompleted;
static SEL sel_blitCommandEncoder;
static SEL sel_endEncoding;
static SEL sel_texture2DDescriptorWithPixelFormat_width_height_mipmapped;
static SEL sel_newTextureWithDescriptor;
static SEL sel_setStorageMode;
static SEL sel_setUsage;
static SEL sel_newBufferWithLength_options;
static SEL sel_contents;
static SEL sel_copyFromTexture_sourceSlice_sourceLevel_sourceOrigin_sourceSize_toBuffer_destinationOffset_destinationBytesPerRow_destinationBytesPerImage;
static SEL sel_addCompletedHandler;
static SEL sel_status;

// Metal classes
static Class MTLTextureDescriptor_class;

// Metal constants
#define MTLPixelFormatBGRA8Unorm 80
#define MTLTextureUsageRenderTarget 0x0004
#define MTLTextureUsageShaderRead 0x0001
#define MTLStorageModePrivate 2
#define MTLStorageModeShared 1
#define MTLResourceStorageModeShared 1
#define MTLCommandBufferStatusCompleted 4

// Internal structures
struct gpu_device {
    id device;
    id command_queue;
};

struct gpu_texture {
    id texture;
    int width;
    int height;
};

struct gpu_readback_buffer {
    id buffer;
    size_t size;
};

struct gpu_command_buffer {
    id cmd_buffer;
    bool completed;
};

// Completion handler for command buffer
typedef void (^CommandBufferHandler)(id);

// Helper to call ObjC methods without arguments
static id msgSend0(id obj, SEL sel) {
    return ((id (*)(id, SEL))objc_msgSend)(obj, sel);
}

// Helper to call ObjC methods with one argument
static id msgSend1(id obj, SEL sel, id arg) {
    return ((id (*)(id, SEL, id))objc_msgSend)(obj, sel, arg);
}

// Helper to call ObjC methods with integer argument
static void msgSendInt(id obj, SEL sel, int arg) {
    ((void (*)(id, SEL, int))objc_msgSend)(obj, sel, arg);
}

// Initialize selectors (called once)
static void init_selectors(void) {
    static bool initialized = false;
    if (initialized) return;

    sel_alloc = sel_registerName("alloc");
    sel_init = sel_registerName("init");
    sel_release = sel_registerName("release");
    sel_retain = sel_registerName("retain");
    sel_newCommandQueue = sel_registerName("newCommandQueue");
    sel_commandBuffer = sel_registerName("commandBuffer");
    sel_commit = sel_registerName("commit");
    sel_waitUntilCompleted = sel_registerName("waitUntilCompleted");
    sel_blitCommandEncoder = sel_registerName("blitCommandEncoder");
    sel_endEncoding = sel_registerName("endEncoding");
    sel_texture2DDescriptorWithPixelFormat_width_height_mipmapped =
        sel_registerName("texture2DDescriptorWithPixelFormat:width:height:mipmapped:");
    sel_newTextureWithDescriptor = sel_registerName("newTextureWithDescriptor:");
    sel_setStorageMode = sel_registerName("setStorageMode:");
    sel_setUsage = sel_registerName("setUsage:");
    sel_newBufferWithLength_options = sel_registerName("newBufferWithLength:options:");
    sel_contents = sel_registerName("contents");
    sel_copyFromTexture_sourceSlice_sourceLevel_sourceOrigin_sourceSize_toBuffer_destinationOffset_destinationBytesPerRow_destinationBytesPerImage =
        sel_registerName("copyFromTexture:sourceSlice:sourceLevel:sourceOrigin:sourceSize:toBuffer:destinationOffset:destinationBytesPerRow:destinationBytesPerImage:");
    sel_addCompletedHandler = sel_registerName("addCompletedHandler:");
    sel_status = sel_registerName("status");

    MTLTextureDescriptor_class = objc_getClass("MTLTextureDescriptor");

    initialized = true;
}

gpu_device_t* gpu_init(void) {
    init_selectors();

    gpu_device_t* device = (gpu_device_t*)malloc(sizeof(gpu_device_t));

    // Create Metal device
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wcast-function-type-mismatch"
    device->device = ((id (*)(void))MTLCreateSystemDefaultDevice)();
#pragma clang diagnostic pop

    if (!device->device) {
        free(device);
        return NULL;
    }
    msgSend0(device->device, sel_retain);

    // Create command queue
    device->command_queue = msgSend0(device->device, sel_newCommandQueue);
    msgSend0(device->command_queue, sel_retain);

    return device;
}

void* gpu_get_native_device(gpu_device_t* device) {
    return device->device;
}

gpu_texture_t* gpu_create_texture(gpu_device_t* device, int width, int height) {
    gpu_texture_t* texture = (gpu_texture_t*)malloc(sizeof(gpu_texture_t));
    texture->width = width;
    texture->height = height;

    // Create texture descriptor
    id tex_desc = ((id (*)(id, SEL, int, int, int, BOOL))objc_msgSend)(
        MTLTextureDescriptor_class,
        sel_texture2DDescriptorWithPixelFormat_width_height_mipmapped,
        MTLPixelFormatBGRA8Unorm, width, height, NO);

    msgSendInt(tex_desc, sel_setStorageMode, MTLStorageModePrivate);
    msgSendInt(tex_desc, sel_setUsage, MTLTextureUsageRenderTarget | MTLTextureUsageShaderRead);

    texture->texture = msgSend1(device->device, sel_newTextureWithDescriptor, tex_desc);
    msgSend0(texture->texture, sel_retain);

    return texture;
}

void* gpu_get_native_texture(gpu_texture_t* texture) {
    return texture->texture;
}

gpu_readback_buffer_t* gpu_create_readback_buffer(gpu_device_t* device, size_t size) {
    gpu_readback_buffer_t* buffer = (gpu_readback_buffer_t*)malloc(sizeof(gpu_readback_buffer_t));
    buffer->size = size;

    buffer->buffer = ((id (*)(id, SEL, size_t, int))objc_msgSend)(
        device->device, sel_newBufferWithLength_options, size, MTLResourceStorageModeShared);
    msgSend0(buffer->buffer, sel_retain);

    return buffer;
}

gpu_command_buffer_t* gpu_readback_texture_async(
    gpu_device_t* device,
    gpu_texture_t* texture,
    gpu_readback_buffer_t* buffer,
    int width,
    int height
) {
    gpu_command_buffer_t* cmd = (gpu_command_buffer_t*)malloc(sizeof(gpu_command_buffer_t));
    cmd->completed = false;

    // Create command buffer
    cmd->cmd_buffer = msgSend0(device->command_queue, sel_commandBuffer);
    msgSend0(cmd->cmd_buffer, sel_retain);

    // Create blit encoder
    id blit_encoder = msgSend0(cmd->cmd_buffer, sel_blitCommandEncoder);

    // Setup copy operation
    typedef struct { NSUInteger x, y, z; } MyMTLOrigin;
    typedef struct { NSUInteger width, height, depth; } MyMTLSize;

    MyMTLOrigin origin = {0, 0, 0};
    MyMTLSize size = {width, height, 1};

    ((void (*)(id, SEL, id, NSUInteger, NSUInteger, MyMTLOrigin, MyMTLSize, id,
               NSUInteger, NSUInteger, NSUInteger))objc_msgSend)(
        blit_encoder,
        sel_copyFromTexture_sourceSlice_sourceLevel_sourceOrigin_sourceSize_toBuffer_destinationOffset_destinationBytesPerRow_destinationBytesPerImage,
        texture->texture, 0, 0, origin, size,
        buffer->buffer, 0, width * 4, width * height * 4);

    msgSend0(blit_encoder, sel_endEncoding);

    // Add completion handler to mark as complete
    CommandBufferHandler handler = Block_copy(^(id cmd_buf) {
        (void)cmd_buf;
        cmd->completed = true;
    });

    ((void (*)(id, SEL, CommandBufferHandler))objc_msgSend)(
        cmd->cmd_buffer, sel_addCompletedHandler, handler);
    Block_release(handler);

    return cmd;
}

void gpu_submit_commands(gpu_command_buffer_t* cmd_buffer, bool wait) {
    msgSend0(cmd_buffer->cmd_buffer, sel_commit);

    if (wait) {
        msgSend0(cmd_buffer->cmd_buffer, sel_waitUntilCompleted);
        cmd_buffer->completed = true;
    }
}

bool gpu_is_readback_complete(gpu_command_buffer_t* cmd_buffer) {
    if (cmd_buffer->completed) return true;

    int status = ((int (*)(id, SEL))objc_msgSend)(cmd_buffer->cmd_buffer, sel_status);
    if (status == MTLCommandBufferStatusCompleted) {
        cmd_buffer->completed = true;
        return true;
    }
    return false;
}

void* gpu_get_readback_data(gpu_readback_buffer_t* buffer) {
    return msgSend0(buffer->buffer, sel_contents);
}

void gpu_copy_readback_data(gpu_readback_buffer_t* buffer, void* dst, size_t size) {
    void* src = msgSend0(buffer->buffer, sel_contents);
    memcpy(dst, src, size < buffer->size ? size : buffer->size);
}

void gpu_destroy_command_buffer(gpu_command_buffer_t* cmd_buffer) {
    if (cmd_buffer) {
        msgSend0(cmd_buffer->cmd_buffer, sel_release);
        free(cmd_buffer);
    }
}

void gpu_destroy_texture(gpu_texture_t* texture) {
    if (texture) {
        msgSend0(texture->texture, sel_release);
        free(texture);
    }
}

void gpu_destroy_readback_buffer(gpu_readback_buffer_t* buffer) {
    if (buffer) {
        msgSend0(buffer->buffer, sel_release);
        free(buffer);
    }
}

void gpu_destroy(gpu_device_t* device) {
    if (device) {
        msgSend0(device->command_queue, sel_release);
        msgSend0(device->device, sel_release);
        free(device);
    }
}